pycyphal/dsdl/_import_hook.py

```diff
@@ -0,0 +1,136 @@
+# Copyright (c) 2019 OpenCyphal
+# This software is distributed under the terms of the MIT License.
+
+import logging
+import sys
+import os
+from typing import Iterable, Optional, Union
+import pathlib
+from . import compile
+
+from importlib.abc import MetaPathFinder
+from importlib.util import spec_from_file_location
+
+_AnyPath = Union[str, pathlib.Path]
+
+_logger = logging.getLogger(__name__)
+
+
+class DsdlMetaFinder(MetaPathFinder):
+    def __init__(
+        self,
+        lookup_directories: Iterable[_AnyPath],
+        output_directory: _AnyPath,
+        allow_unregulated_fixed_port_id: bool,
+    ) -> None:
+        super().__init__()
+
+        _logger.debug("lookup dirs: %s", lookup_directories)
+        _logger.debug("output dir: %s", output_directory)
+
+        self.lookup_directories = list(map(str, lookup_directories))
+        self.output_directory = output_directory
+        self.allow_unregulated_fixed_port_id = allow_unregulated_fixed_port_id
+        self.root_namespace_directories: Iterable[pathlib.Path] = list()
+
+        # Build a list of root namespace directories from lookup directories.
+        # Any dir inside any of the lookup directories is considered a root namespace.
+        for dir in self.lookup_directories:
+            for namespace in pathlib.Path(dir).iterdir():
+                if namespace.is_dir():
+                    _logger.debug("Using root namespace %s at %s", namespace.name, namespace)
+                    self.root_namespace_directories.append(namespace)
+
+    def find_source_dir(self, root_namespace: str) -> Optional[pathlib.Path]:
+        """
+        Finds DSDL source directory for a given root namespace name.
+        """
+        for namespace_dir in self.root_namespace_directories:
+            if namespace_dir.name == root_namespace:
+                return namespace_dir
+        return None
+
+    def is_compiled(self, root_namespace: str) -> bool:
+        """
+        Returns true if given root namespace exists in output directory (compiled).
+        """
+        return pathlib.Path(self.output_directory, root_namespace).exists()
+
+    def find_spec(self, fullname, path, target=None):
+        _logger.debug("Attempting to load module %s as DSDL", fullname)
+
+        # DSDL root namespace is equal to module name
+        root_namespace = fullname
```

Comment:
If my namespace is named `if` the module name would be `if_`, this would break. [`pycyphal.dsdl.get_attribute()`](https://github.com/OpenCyphal/pycyphal/blob/338e206b375c230b96411ca4d5249087273bce36/pycyphal/dsdl/_composite_object.py#L146-L158) has a retry to handle this.